keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1285 @@
+of the number of pod failures for a given index. It is set to `0` for the first
+pod created for a given index.
+
+When Job controller sees a failed pod corresponding to a given index, and the
+value of the annotation `batch.kubernetes.io/job-index-failure-count` is greater
+or equal to the configured backoff limit per index then the index is marked
+as failed and added to `.status.failedIndexes`.
+
+When Job controller creates replacement pods for failed pods for a given
+index it checks if the index isn't finished yet (it is not in
+`.status.failedIndexes` nor `.status.completedIndexes`).
+Then, if `x` is the highest `batch.kubernetes.io/job-index-failure-count`
+for the index, the newly created pod will have the annotation set to `x+1`.
+
+In order to keep track of the number of failures per index, the Job controller
+removes finalizers of a failed pod for a given index, only once the replacement
+pod (with incremented value of `batch.kubernetes.io/job-index-failure-count`) is
+created, or the index is marked as failed in `.status.failedIndexes`. This means
+that these are the main steps when handling a failed pod to prepare it for
+deletion:
+1. Pod is recognized as failed
+2. pod UID is recorded in Job status (`.status.uncountedTerminatedPods`)
+3. the replacement Pod is created (the new condition)
+4. Pod's finalizer is removed
+
+So, the new feature adds a dependency between steps (3.) and (4.) as previously
+these steps could be performed in any order.
+
+Note that, the first step (1.) will also be impacted by
+[KEP-3939: Consider Terminating pods as active pods in Jobs.](https://github.com/kubernetes/enhancements/issues/3939)
+
+### Failed indexes format
+
+The format of the `.status.failedIndexes` field is analogous to the one used for
+successful indexes, represented by the `completedIndexes` field, which is a
+text format grouping consecutive integers into ranges. In a special case, when
+the indexes are non-consecutive they are represented by comma-separated numbers.
+In the worst-case scenario this is a string of comma-separated even values from
+`0` to `.spec.completions-1`. In order to constrain the size of the field we
+cap the number of completions (see [The Job object too big](#the-job-object-too-big)
+for more details).
+
+### Job completion
+
+When backoff limit per index is used, then we execute indexes until all of them
+are completed (either failed or succeeded).
+
+Then, the Job is marked as completed (the `Complete` Job condition type) when
+all indexes are succeeded. The Job is marked as failed (the `Failed` Job condition)
+when at least one index is failed. The `Failed` condition is added once
+all indexes completed their execution (either failed or succeeded), or when
+the number of failed indexes exceeds the specified `.spec.runPolicy.maxFailedIndexes`.
+The `reason` field for the failed Job is `FailedIndexes` and the `message` field
+will list up to a couple of the failed indexes.
```

Comment:
what's "a couple"? Please be more precise: 2, 3, 10?

Reply:
I was thinking this is implementation detail which can be figured out later. I think 3 sounds sensible, will update.

Reply:
I just find the word "couple" very confusing, as I'm not sure if you mean literally 2.

Reply:
updated